Ska/engarchive/derived/comps.py

```diff
@@ -532,3 +534,205 @@ def get_msid_attrs(self, tstart, tstop, msid, msid_args):
         }
 
         return out
+
+
+@functools.lru_cache(maxsize=1)
+def get_roll_pitch_tlm(start, stop):
+    from .. import fetch
+
+    msids = ["6sares1", "6sares2", "6sunsa1", "6sunsa2", "6sunsa3"]
+    dat = fetch.MSIDset(msids, start, stop)
+
+    # Filter out of range values. This happens just at the beginning of safe mode.
+    for msid in msids:
+        if "6sares" in msid:
+            x0, x1 = 0, 180
+        else:
+            x0, x1 = -1, 1
+        dat[msid].bads |= (dat[msid].vals < x0) | (dat[msid].vals > x1)
+
+    dat.interpolate(times=dat[msids[0]].times, bad_union=True)
+    return dat
+
+
+def calc_css_pitch_safe(start, stop):
+    from .pcad import arccos_clip
+
+    # Get the raw telemetry value in user-requested unit system
+    dat = get_roll_pitch_tlm(start, stop)
+
+    sa_ang_avg = (1.0 * dat["6sares1"].vals + 1.0 * dat["6sares2"].vals) / 2
+    sinang = np.sin(np.radians(sa_ang_avg))
+    cosang = np.cos(np.radians(sa_ang_avg))
+    # Rotate CSS sun vector from SA to ACA frame
+    css_aca = np.array(
+        [
+            sinang * dat["6sunsa1"].vals - cosang * dat["6sunsa3"].vals,
+            dat["6sunsa2"].vals * 1.0,
+            cosang * dat["6sunsa1"].vals + sinang * dat["6sunsa3"].vals,
+        ]
+    )
+    # Normalize sun vec (again) and compute pitch
+    magnitude = np.sqrt((css_aca * css_aca).sum(axis=0))
+    magnitude[magnitude == 0.0] = 1.0
+    sun_vec_norm = css_aca / magnitude
+    vals = np.degrees(arccos_clip(sun_vec_norm[0]))
+
+    return dat.times, vals
+
+
+def calc_css_roll_safe(start, stop):
+    """Off-Nominal Roll Angle from CSS Data in ACA Frame [Deg]
+
+    Defined as the rotation about the ACA X-axis required to align the sun
+    vector with the ACA X/Z plane.
+
+    Calculated by rotating the CSS sun vector from the SA-1 frame to ACA frame
+    based on the solar array angles 6SARES1 and 6SARES2.
+
+    """
+    # Get the raw telemetry value in user-requested unit system
+    dat = get_roll_pitch_tlm(start, stop)
+
+    sa_ang_avg = (dat["6sares1"].vals + dat["6sares2"].vals) / 2
+    sinang = np.sin(np.radians(sa_ang_avg))
+    cosang = np.cos(np.radians(sa_ang_avg))
+    # Rotate CSS sun vector from SA to ACA frame
+    css_aca = np.array(
+        [
+            sinang * dat["6sunsa1"].vals - cosang * dat["6sunsa3"].vals,
+            dat["6sunsa2"].vals,
+            cosang * dat["6sunsa1"].vals + sinang * dat["6sunsa3"].vals,
+        ]
+    )
+    # Normalize sun vec (again) and compute pitch
+    magnitude = np.sqrt((css_aca * css_aca).sum(axis=0))
+    # Don't exactly know why zero values can happen, but it does, so just drop those.
+    ok = magnitude > 0
```

Comment:
Not entirely sure about the different implementation here for the same issue?

Reply:
Fixed by 3936f79.